### PR TITLE
Added gamemode to CondIsInvulnerable.java

### PR DESCRIPTION
**Pull request by JakeGBLP**
### Description
Added gamemode to ConsIsInvulnerable

---
**Target Minecraft Versions:** 1.20.6
**Requirements:** Paper
**Related Issues:** none

### CODE REVIEW

File: src/main/java/ch/njol/skript/conditions/CondIsInvulnerable.java [hunk 1]
```diff
@@ -27,18 +28,25 @@
...
-@Description("Checks whether an entity is invulnerable.")
-@Examples("target entity is invulnerable")
-@Since("2.5")
-public class CondIsInvulnerable extends PropertyCondition<Entity> {
+@Description("Checks whether an entity or a gamemode is invulnerable.")
+@Examples({"target entity is invulnerable",
+"loop {_gamemodes::*}:\n"+
+"\tif loop-value is not invulnerable:\n" +
+"\t\tbroadcast \"the gamemode %loop-value% is vulnerable!\""})
+@Since("2.5") // ?
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/conditions/CondIsInvulnerable.java:36]:
```suggestion
@Since("2.5, INSERT VERSION (gamemodes)")
@RequiredPlugins("Paper 1.20.6+ (gamemodes)")
```

File: src/main/java/ch/njol/skript/conditions/CondIsInvulnerable.java [hunk 2]
```diff
@@ -27,18 +28,25 @@
...
 @Name("Is Invulnerable")
-@Description("Checks whether an entity is invulnerable.")
-@Examples("target entity is invulnerable")
-@Since("2.5")
-public class CondIsInvulnerable extends PropertyCondition<Entity> {
+@Description("Checks whether an entity or a gamemode is invulnerable.")
+@Examples({"target entity is invulnerable",
+"loop {_gamemodes::*}:\n"+
+"\tif loop-value is not invulnerable:\n" +
+"\t\tbroadcast \"the gamemode %loop-value% is vulnerable!\""})
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/conditions/CondIsInvulnerable.java:35]:
```suggestion
@Examples({
	"target entity is invulnerable",
	"",
	"loop {_gamemodes::*}:\n"+
		"\tif loop-value is not invulnerable:\n" +
			"\t\tbroadcast \"the gamemode %loop-value% is vulnerable!\""
})
```

File: src/main/java/ch/njol/skript/conditions/CondIsInvulnerable.java [hunk 3]
```diff
@@ -27,18 +28,25 @@
...
+@Since("2.5") // ?
+public class CondIsInvulnerable extends PropertyCondition<Object> {
 	
 	static {
-		register(CondIsInvulnerable.class, PropertyType.BE, "invulnerable", "entities");
+		register(CondIsInvulnerable.class, PropertyType.BE, "invulnerable", "objects");
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/conditions/CondIsInvulnerable.java:40]:
```suggestion
		register(CondIsInvulnerable.class, PropertyType.BE, "invulnerable", "entities/gamemodes");
```

Comment (sovdeeth) [reply src/main/java/ch/njol/skript/conditions/CondIsInvulnerable.java:40]: Actually this needs to check whether the method exists with Skript.methodExists() before adding `gamemodes` to the syntax.

**Review by sovdeeth — CHANGES_REQUESTED** [review]

**Review by sovdeeth — APPROVED** [review]

File: src/main/java/ch/njol/skript/conditions/CondIsInvulnerable.java [hunk 4]
```diff
@@ -18,27 +18,40 @@
...
 @Name("Is Invulnerable")
-@Description("Checks whether an entity is invulnerable.")
-@Examples("target entity is invulnerable")
-@Since("2.5")
-public class CondIsInvulnerable extends PropertyCondition<Entity> {
+@Description("Checks whether an entity or a gamemode is invulnerable.\nFor gamemodes, Paper and Minecraft 1.20.6 are required")
+@Examples({
+	"target entity is invulnerable",
+	"",
+	"loop {_gamemodes::*}:",
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/conditions/CondIsInvulnerable.java:34]:
```suggestion
	"loop all gamemodes:",
```
I believe this should be possible

File: src/main/java/ch/njol/skript/conditions/CondIsInvulnerable.java [hunk 5]
```diff
@@ -18,27 +18,40 @@
...
+	"target entity is invulnerable",
+	"",
+	"loop all gamemodes:",
+		"\tif loop-value is not invulnerable:",
+			"\t\tbroadcast \"the gamemode %loop-value% is vulnerable!\""
+})
+@Since("2.5, INSERT VERSION (gamemode)")
+@RequiredPlugins("Paper 1.20.6+ (gamemodes)")
+public class CondIsInvulnerable extends PropertyCondition<Object> {
+	private static final boolean GAMEMODE = Skript.methodExists(GameMode.class,"isInvulnerable");
```
Comment (Pikachu920) [inline src/main/java/ch/njol/skript/conditions/CondIsInvulnerable.java:41]: missing a space after the comma

Comment (Pikachu920) [reply src/main/java/ch/njol/skript/conditions/CondIsInvulnerable.java:41]: can we also name this `SUPPORTS_GAMEMODE` or something besides just gamemode

File: src/test/skript/tests/syntaxes/conditions/CondIsInvulnerable.sk [hunk 6]
```diff
@@ -0,0 +1,3 @@
+test "CondIsInvulnerable":
+  assert creative is invulnerable with "Creative mode should be invulnerable"
+  assert survival is not invulnerable with "Survival mode should not be invulnerable"
```
Comment (sovdeeth) [inline src/test/skript/tests/syntaxes/conditions/CondIsInvulnerable.sk:3]:
```suggestion
  assert survival is not invulnerable with "Survival mode should not be invulnerable"

```
this should also test the cases for entities imo, and the cases where the input is null (`{_}`)

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/test/skript/tests/syntaxes/conditions/CondIsInvulnerable.sk (hunk 6 above)
Comment (sovdeeth) [reply src/test/skript/tests/syntaxes/conditions/CondIsInvulnerable.sk:3]: you also need a `running minecraft "1.20.6"` condition

Comment (Pikachu920) [reply src/test/skript/tests/syntaxes/conditions/CondIsInvulnerable.sk:3]: you can actually use the method exists condition!

File: src/test/skript/tests/syntaxes/conditions/CondIsInvulnerable.sk [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+
+test "CondIsInvulnerable":
+	spawn armor stand at spawn of world "world":
+		make entity invulnerable
+		assert entity is invulnerable with "Entity should be invulnerable"
+		make entity vulnerable
+		assert entity is vulnerable with "Entity should be vulnerable"
+		delete entity
+	set {_} to {_}
+	assert {_} is vulnerable with "Empty variable should be vulnerable"
```
Comment (sovdeeth) [inline src/test/skript/tests/syntaxes/conditions/CondIsInvulnerable.sk:13]:
```suggestion
	assert {_} is vulnerable with "Empty variable should be vulnerable"

```

**Review by sovdeeth — APPROVED** [review]

File: src/test/skript/tests/syntaxes/conditions/CondIsInvulnerable.sk (hunk 7 above)
Comment (UnderscoreTud) [inline src/test/skript/tests/syntaxes/conditions/CondIsInvulnerable.sk:13]:
```suggestion
	assert {_none} is vulnerable with "Empty variable should be vulnerable"
```

**Review by UnderscoreTud — APPROVED** [review]

**Review by Moderocky — APPROVED** [review]

**Comment by sovdeeth** [comment]
tests failing due to invalid syntax

File: src/test/skript/tests/syntaxes/conditions/CondIsInvulnerable.sk [hunk 8]
```diff
@@ -0,0 +1,12 @@
+test "CondIsInvulnerable" when running minecraft "1.20.6":
+	assert creative is invulnerable with "Creative mode should be invulnerable"
+	assert survival is not invulnerable with "Survival mode should not be invulnerable"
+
+test "CondIsInvulnerable":
+	spawn armor stand at spawn of world "world":
+		make entity invulnerable
+		assert entity is invulnerable with "Entity should be invulnerable"
+		make entity vulnerable
+		assert entity is vulnerable with "Entity should be vulnerable"
```
Comment (sovdeeth) [inline src/test/skript/tests/syntaxes/conditions/CondIsInvulnerable.sk:10]: invalid

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/test/skript/tests/syntaxes/conditions/CondIsInvulnerable.sk [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+	assert survival is not invulnerable with "Survival mode should not be invulnerable"
+
+test "CondIsInvulnerable":
+	spawn armor stand at spawn of world "world":
+		make entity invulnerable
+		assert entity is invulnerable with "Entity should be invulnerable"
+		make entity vulnerable
+		assert entity is not invulnerable with "Entity should be vulnerable"
+		delete entity
+	assert {_none} is vulnerable with "Empty variable should be vulnerable"
```
Comment (sovdeeth) [inline src/test/skript/tests/syntaxes/conditions/CondIsInvulnerable.sk:12]: still invalid

**Review by sovdeeth — CHANGES_REQUESTED** [review]

**Review by sovdeeth — APPROVED** [review]
